private/bufpkg/bufconfig/write.go

```diff
@@ -273,30 +257,173 @@ func writeConfig(
+	breakingConfig := writeConfigOptions.breakingConfig
+	if breakingConfig != nil {
+		breakingConfigVersion = breakingConfig.Version
+		if breakingConfigVersion != version {
+			return fmt.Errorf("invalid version found for breaking config: %q", breakingConfigVersion)
+		}
+		config.Breaking = breakingConfig
 	}
-	for _, dependencyModuleReference := range writeConfigOptions.dependencyModuleReferences {
-		externalConfigV1.Deps = append(
-			externalConfigV1.Deps,
-			dependencyModuleReference.String(),
-		)
+	var lintConfigVersion string
+	lintConfig := writeConfigOptions.lintConfig
+	if lintConfig != nil {
+		lintConfigVersion = lintConfig.Version
+		if lintConfigVersion != version {
+			return fmt.Errorf("invalid version found for lint config: %q", lintConfigVersion)
+		}
+		config.Lint = lintConfig
+	}
+	// We should never hit this condition since we are already validating against `version`.
+	if breakingConfigVersion != lintConfigVersion {
+		return fmt.Errorf("breaking config version %q does not match lint config version %q", breakingConfigVersion, lintConfigVersion)
+	}
+	var dependencies []string
+	if len(writeConfigOptions.dependencyModuleReferences) > 0 {
+		dependencies = make([]string, len(writeConfigOptions.dependencyModuleReferences))
+		for i, dependencyModuleReference := range writeConfigOptions.dependencyModuleReferences {
+			dependencies[i] = dependencyModuleReference.String()
+		}
 	}
-	tmplData := tmplUndocumentedData
 	if writeConfigOptions.documentationComments {
-		tmplData = tmplDocumentationCommentsData
+		// Write out config using template with document comments.
+		return writeCommentedConfig(ctx, writeBucket, config, dependencies, writeConfigOptions.uncomment)
+	}
+	// Write out raw default config without comments using externalConfig{V1, V1Beta1}
+	return writeExternalConfig(ctx, writeBucket, version, config, dependencies)
+}
+
+func writeExternalConfig(
+	ctx context.Context,
+	writeBucket storage.WriteBucket,
+	version string,
+	config *Config,
+	dependencies []string,
+) error {
+	var name string
+	if config.ModuleIdentity != nil {
+		name = config.ModuleIdentity.IdentityString()
+	}
+	breakingConfig := config.Breaking
+	lintConfig := config.Lint
+	switch version {
+	case V1Beta1Version:
+		var externalBreakingConfig bufbreakingconfig.ExternalConfigV1Beta1
+		if breakingConfig != nil {
+			externalBreakingConfig = bufbreakingconfig.ExternalConfigV1Beta1ForConfig(breakingConfig)
+		}
+		var externalLintConfig buflintconfig.ExternalConfigV1Beta1
+		if lintConfig != nil {
+			externalLintConfig = buflintconfig.ExternalConfigV1Beta1ForConfig(lintConfig)
+		}
+		externalConfig := ExternalConfigV1Beta1{
+			Name:     name,
+			Version:  config.Version,
+			Deps:     dependencies,
+			Breaking: externalBreakingConfig,
+			Lint:     externalLintConfig,
+		}
+		var buffer bytes.Buffer
+		encoder := yaml.NewEncoder(&buffer)
+		encoder.SetIndent(2)
+		if err := encoder.Encode(externalConfig); err != nil {
+			return err
+		}
```

Comment:
Discussed offline - 

This will produce non-deterministic results because the YAML structure contains a `map[string][]string` nested within it.

Right now, this isn't actually user-facing - the only time we actually write these potentially non-deterministic values is in the users' local module cache. It's definitely not ideal, but we can follow-up with a fix after this PR is merged (i.e. using the `yaml.Node` API in `yaml.v3`).

What matters most is that the digest algorithm is deterministic and the `buf.yaml` can be successfully round-tripped without losing any information (just sometimes in a different order, such as with `ignore_only`).

We will need a fix in these representations before encoding the `buf.yaml` is formally complete.

Reply:
+1, thanks for comment :) I am currently working on summarizing the requirements for having the config be round-trippable in a stand-alone issue, does it makes sense to bundle that as a strict requirement for the encoding `buf.yaml` issue or should it be its own thing?